File: frontend-next/src/types/roadmap.ts
export interface RoadmapDimension {
  name: string;
  score: number;
  maxScore: number;
  level: string;
}

export interface RoadmapRecommendRequest {
  dimensions: RoadmapDimension[];
  overallLevel: string;
  totalScore: number;
  maxScore: number;
}

export interface RoadmapCourse {
  course: string;
  priority: number;
  reason: string;
  estimatedWeeks: number;
}

export interface RoadmapRecommendResponse {
  success: boolean;
  roadmap: RoadmapCourse[];
  summary: string;
  diagnostic: {
    overallLevel: string;
    totalScore: number;
    maxScore: number;
  };
  error?: string;
}

export function dimensionsToRoadmapInput(
  dimensions: RoadmapDimension[],
  overallLevel: string,
  totalScore: number,
  maxScore: number,
): RoadmapRecommendRequest {
  return {
    dimensions,
    overallLevel,
    totalScore,
    maxScore,
  };
}

File: backend/app/schemas/roadmap.py
from __future__ import annotations

from pydantic import BaseModel, Field


class RoadmapDimension(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    score: int = Field(ge=0, le=100)
    maxScore: int = Field(ge=1, le=100)
    level: str = Field(min_length=1, max_length=20)


class RoadmapRecommendRequest(BaseModel):
    dimensions: list[RoadmapDimension] = Field(min_length=1)
    overallLevel: str = Field(min_length=1, max_length=20)
    totalScore: int = Field(ge=0)
    maxScore: int = Field(ge=1)


class RoadmapCourse(BaseModel):
    course: str
    priority: int
    reason: str
    estimatedWeeks: int


class RoadmapRecommendResponse(BaseModel):
    success: bool
    roadmap: list[RoadmapCourse] = []
    summary: str = ""
    diagnostic: dict = {}
    error: str | None = None
